### DISCUSSION TITLE
what's the limit of curl .

### DISCUSSION THREAD

**Discussion by babA409866** [curl/curl#11069] [Q&A] [answered]
when i read the source code .
CURL_SOCKET_HASH_TABLE_SIZE was set to 911.
If i want to improve the concurrency of curl and need to manage more socket connections simultaneously, do i need to increase the size of CURL_SOCKET_HASH_TABLE_SIZE.
or i just use C++11 thread . each thread create a multi-curl to reach the high concurrency target.
What is the optimal number of easy curls per multi curl for best performance?
when i have to do a lots of requests . others told curl is not the choice . but i think there are some ways to increate the performance.

**Comment by bagder** [comment, marked as answer]
> do i need to increase the size of CURL_SOCKET_HASH_TABLE_SIZE.

Maybe. If you can make a test case where you see noticeably better results by increasing that number, sure. I would expect that there are other/better things to adjust as well or instead. But either way: testing and actual numbers is key for any such (proposed) change.

>  i just use C++11 thread . each thread create a multi-curl

If your transfers are CPU-bound, then you need to use more threads to scale higher. Yes. libcurl currently only runs in a single thread on its own.

>  What is the optimal number of easy curls per multi curl for best performance?

I don't think that can be expressed with a number or a formula. It depends very much on your use case and platform etc. I know users in the past who successfully used thousands.

> others told curl is not the choice

There will always be others out there who knows better. At least they say and think they do. Again, testing and doing measurements is the key. I don't know of any particular architectural limitations in libcurl that would prevent it to scale really well.

I'm also convinced that there are things in the code that can be improved - there always is!
